bundles/org.openhab.core.io.rest.update/src/main/java/org/openhab/core/io/rest/update/internal/restresource/UpdaterRestResource.java

```diff
@@ -0,0 +1,171 @@
+
+    private @Nullable BaseUpdater updater;
+
+    @Activate
+    public UpdaterRestResource() {
+        this.updater = UpdaterFactory.newUpdater();
+    }
+
+    /**
+     * This method handles HTTP GET on the updater resource's 'status' path. It serves a JSON DTO containing the updater
+     * status. This is a read only function so both USER and ADMIN roles are allowed.
+     *
+     * @return HTTP 200 OK, or HTTP 500 INTERNAL SERVER ERROR if the updater is null.
+     */
+    @GET
+    @RolesAllowed({ Role.USER, Role.ADMIN })
+    @Path(UpdaterRestResource.URI_STATUS)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Operation(operationId = "status", summary = "Get the updater status.", responses = {
+            @ApiResponse(responseCode = "200", description = OK_SUCCESS, content = @Content(schema = @Schema(implementation = UpdaterStatusDTO.class))),
+            @ApiResponse(responseCode = "500", description = INT_SERV_ERR_UPDATER_NULL) })
+    public Response status() {
+        BaseUpdater updater = this.updater;
+
+        // return server error if updater is null
+        if (updater == null) {
+            return Response.status(Status.INTERNAL_SERVER_ERROR).entity(INT_SERV_ERR_UPDATER_NULL).build();
+        }
+
+        // populate and return the DTO
+        return Response.ok(updater.getStatusDTO()).build();
+    }
+
+    /**
+     * This method handles HTTP POST on the updater resource's 'execute' path. It accepts a JSON DTO, and if the
+     * parameters are OK starts the updater's update process. It serves a PLAIN TEXT response indicating success or
+     * failure. This is an administrator function so only ADMIN role is allowed.
+     *
+     * @return HTTP 200 OK, HTTP 400 BAD REQUEST if the JSON DTO parameters have errors, or HTTP 500 INTERNAL SERVER
+     *         ERROR if the updater is null.
+     */
+    @POST
+    @RolesAllowed({ Role.ADMIN })
+    @Path(UpdaterRestResource.URI_EXECUTE)
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.TEXT_PLAIN)
+    @Operation(operationId = "execute", summary = "Start updating openHAB to a newer version.", responses = {
+            @ApiResponse(responseCode = "200", description = OK_SUCCESS),
+            @ApiResponse(responseCode = "400", description = BAD_REQUEST),
+            @ApiResponse(responseCode = "500", description = INT_SERV_ERR_UPDATER_NULL) })
+    public Response execute(
+            @Parameter(description = "Update command settings", required = true) @Valid UpdaterExecuteDTO dto) {
+        BaseUpdater updater = this.updater;
+
+        // return server error if updater is null
+        if (updater == null) {
+            return Response.status(Status.INTERNAL_SERVER_ERROR).entity(INT_SERV_ERR_UPDATER_NULL).build();
+        }
+
+        // check targetNewVersionType
+        if (dto.targetNewVersionType != null) {
+            try {
+                updater.setNewVersionType(dto.targetNewVersionType);
```

Comment:
Having a single setter method for all three properties would also be weird (IMHO). The incoming DTO contains three properties, each of which might be invalid (or absent). So I prefer to use a setter for each property in turn, so that the respective setter can do the validation checks appropriate to itself. So a) target version is validated to match STABLE/MILESTONE/SNAPSHOT, b) user name is validated to be alphanumeric with no white space, and c) password etc.